File: backend/patient-service/app/redis/redis_service.py
from redis.asyncio import Redis

from app.common.logger import get_logger
from app.redis.server import redis_client

logger = get_logger("RedisService")


class RedisService:

    def __init__(self):
        self.redis: Redis = redis_client

    # * Ping Redis
    async def ping(self) -> bool:
        connected = await self.redis.ping()

        if connected:
            logger.info("Redis connected successfully")
        else:
            logger.error("Redis connection failed")

        return bool(connected)

    # * Get key-value from Redis
    async def get(self, key: str) -> str | None:
        value = await self.redis.get(key)

        if isinstance(value, bytes):
            return value.decode()

        return value

    # * Set key-value from Redis
    async def set(
        self,
        key: str,
        value: str,
        ttl: int | None = None,
    ) -> None:
        if ttl is not None:
            await self.redis.set(
                key,
                value,
                ex=ttl,
            )
            return

        await self.redis.set(key, value)

    # * Delete key from redis
    async def delete(self, key: str) -> None:
        await self.redis.delete(key)

    # * Checking Existence of key in Redis
    async def exists(self, key: str) -> bool:
        return bool(await self.redis.exists(key))

    # * Returns the time-to-live of a key in redis
    async def ttl(self, key: str) -> int:
        return await self.redis.ttl(key)

    # * Setting key with time-to-live
    async def set_with_ttl(
        self,
        key: str,
        value: str,
        ttl: int,
    ) -> None:
        await self.redis.set(
            key,
            value,
            ex=ttl,
        )

    # * Returning redis pipeline
    def pipeline(self):
        return self.redis.pipeline()

    # * Set Distribution lock
    async def acquire_lock(
        self,
        key: str,
        token: str,
        ttl: int = 10,
    ) -> bool:
        result = await self.redis.set(
            key,
            token,
            nx=True,
            ex=ttl,
        )

        return result is True

    # * Release Distribution Lock
    async def release_lock(
        self,
        key: str,
        token: str,
    ) -> bool:
        script = """
        if redis.call("GET", KEYS[1]) == ARGV[1] then
            return redis.call("DEL", KEYS[1])
        else
            return 0
        end
        """

        result = await self.redis.eval(
            script,
            1,
            key,
            token,
        )

        return result == 1

    # * Close Redis Connection
    async def close(self) -> None:
        await self.redis.aclose()
        logger.info("Redis connection closed gracefully")

File: backend/health-institute-service/src/redis/redis.service.ts
import {
  Injectable,
  Inject,
  OnModuleInit,
  OnModuleDestroy,
  Logger,
} from '@nestjs/common';
import Redis from 'ioredis';

import { REDIS_CLIENT } from '../common/utils/constants';

@Injectable()
export class RedisService implements OnModuleInit, OnModuleDestroy {
  private readonly logger = new Logger(RedisService.name);

  constructor(@Inject(REDIS_CLIENT) private readonly redis: Redis) {}

  onModuleInit() {
    this.redis.on('connect', () => {
      this.logger.log('Redis connecting');
    });

    if (this.redis.status !== 'ready' && this.redis.status !== 'connecting') {
      void this.redis.connect();
    }

    this.redis.on('ready', () => {
      this.logger.log('Redis ready');
    });

    this.redis.on('error', (error: Error) => {
      this.logger.error(error.message, error.stack);
    });

    this.redis.on('close', () => {
      this.logger.warn('Redis connection closed');
    });

    this.redis.on('reconnecting', (delay: number) => {
      this.logger.warn(`Redis reconnecting, delay=${delay}ms`);
    });

    this.redis.on('end', () => {
      this.logger.warn('Redis connection ended');
    });
  }

  getClient(): Redis {
    return this.redis;
  }

  pipeline() {
    return this.redis.pipeline();
  }

  async get(key: string): Promise<string | null> {
    return this.redis.get(key);
  }

  async set(key: string, value: string, ttl?: number): Promise<void> {
    if (ttl !== undefined) {
      await this.redis.set(key, value, 'EX', ttl);
      return;
    }

    await this.redis.set(key, value);
  }

  async increment(key: string): Promise<number> {
    return this.redis.incr(key);
  }

  async decrement(key: string): Promise<number> {
    return this.redis.decr(key);
  }

  async expire(key: string, seconds: number): Promise<void> {
    await this.redis.expire(key, seconds);
  }

  async delete(key: string): Promise<void> {
    await this.redis.del(key);
  }

  async exists(key: string): Promise<boolean> {
    return (await this.redis.exists(key)) === 1;
  }

  async ttl(key: string): Promise<number> {
    return this.redis.ttl(key);
  }

  async setWithTTL(key: string, value: string, ttl: number): Promise<void> {
    await this.redis.set(key, value, 'EX', ttl);
  }

  async acquireLock(key: string, token: string, ttl: number): Promise<boolean> {
    const result = await this.redis.set(key, token, 'EX', ttl, 'NX');

    return result === 'OK';
  }

  async releaseLock(key: string, token: string): Promise<boolean> {
    const script = `
      if redis.call("GET", KEYS[1]) == ARGV[1] then
        return redis.call("DEL", KEYS[1])
      else
        return 0
      end
    `;

    const result = await this.redis.eval(script, 1, key, token);

    return result === 1;
  }

  async sleep(ms: number): Promise<void> {
    await new Promise((resolve) => setTimeout(resolve, ms));
  }

  async onModuleDestroy() {
    try {
      await this.redis.quit();
      this.logger.log('Redis connection closed gracefully');
    } catch (error) {
      this.logger.error('Error closing Redis', error as Error);

      this.redis.disconnect();
    }
  }
}
